fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBIndexOutput.java

```diff
@@ -0,0 +1,150 @@
+    private int currentSize = 0;
+    private ByteBuffer buffer;
+    private final String resourceDescription;
+    private final FDBDirectory fdbDirectory;
+    private final long blockSize;
+    private final CRC32 crc;
+    private final long id;
+    public static final ArrayBlockingQueue<ByteBuffer> BUFFERS;
+    public static final int POOL_SIZE = 100;
+
+    static {
+        BUFFERS = new ArrayBlockingQueue<>(POOL_SIZE);
+        for (int i = 0; i < POOL_SIZE; i++) {
+            BUFFERS.add(ByteBuffer.allocate(FDBDirectory.DEFAULT_BLOCK_SIZE));
+        }
+    }
+
+    /**
+     * Create an FDBIndexOutput given a resource description, name, and FDBDirectory.
+     *
+     * @param resourceDescription description of resource
+     * @param name name of resource
+     * @param fdbDirectory existing FDBDirectory
+     */
+    public FDBIndexOutput(String resourceDescription, String name, FDBDirectory fdbDirectory) {
+        super(resourceDescription, name);
+        LOG.trace("init() -> resource={}, name={}", resourceDescription, name);
+        this.resourceDescription = resourceDescription;
+        this.fdbDirectory = fdbDirectory;
+        blockSize = fdbDirectory.getBlockSize();
+        buffer = BUFFERS.poll();
+        if (buffer == null) {
+            buffer = ByteBuffer.allocate((int)blockSize);
+        }
+        crc = new CRC32();
+        id = fdbDirectory.getIncrement();
+    }
+
+    /**
+     * Close the directory which writes the FileReference.
+     */
+    @Override
+    @SpotBugsSuppressWarnings(value = "RV_RETURN_VALUE_IGNORED_BAD_PRACTICE", justification = "it is fine if it is not accepted")
+    public void close() {
+        LOG.trace("close() -> resource={}", resourceDescription);
+        flush();
+        fdbDirectory.writeFDBLuceneFileReference(resourceDescription, new FDBLuceneFileReference(id, currentSize, blockSize));
+        BUFFERS.offer(buffer);
+    }
+
+    @Override
+    public long getFilePointer() {
+        LOG.trace("getFilePointer() -> resource={}, pointer={}", resourceDescription, currentSize);
+        return currentSize;
+    }
+
+    @Override
+    public long getChecksum() {
+        LOG.trace("getChecksum() -> resource={}, checksum={}", resourceDescription, crc.getValue());
+        return crc.getValue();
+    }
+
+    @Override
+    public void writeByte(final byte b) {
+        LOG.trace("writeByte() -> resource={}", resourceDescription);
+        buffer.put(b);
+        crc.update(b);
+        currentSize++;
+        if (currentSize % blockSize == 0) {
+            flush();
+        }
+    }
+
+    @Override
+    public void copyBytes(final DataInput input, final long numBytes) throws IOException {
+        LOG.debug("copy bytes input={}, numbytes={}", input, numBytes);
+        super.copyBytes(input, numBytes);
+    }
+
+    @Override
+    public void writeBytes(final byte[] bytes, final int offset, final int length) {
```

Comment:
I am not familiar with the particulars of Lucene APIs, but I haven't seen anything that would "truncate" a file - reduce the size of the file (that would necessitate clearing blocks at the end).

Reply:
Check out deleteFile.  Since Lucene files are immutable, you see a lot of merges and then deletes.

`   /**
     * deletes the file reference under the provided name.
     * @param name the name for the file reference
     */
    @Override
    public void deleteFile(String name) throws IOException {
        try {
            LOG.trace("deleteFile -> {}", name);
            hasWritten = true;
            getFDBLuceneFileReference(name).thenAcceptAsync(
                    (value) -> {
                        if (value == null) {
                            throw new RuntimeException(new NoSuchFileException(name));
                        }
                        txn.clear(metaSubspace.pack(name));
                        txn.clear(dataSubspace.subspace(Tuple.from(value.getId())).range());
                        this.fileReferenceCache.invalidate(name);
                    }
            ).join();
        } catch (CompletionException completionException) {
            if (completionException.getCause() instanceof RuntimeException &&
                    completionException.getCause().getCause() instanceof NoSuchFileException) {
                // This is here to mimic the expectation from existing file system implementations in Lucene
                throw (NoSuchFileException) completionException.getCause().getCause();
            }
            throw completionException;
        }
    }`